fpdf/template.py

```diff
@@ -60,80 +38,105 @@ def __init__(
+            return float((s.strip() or default).replace(decimal_sep, "."))
+
+        key_config = (
+            # key, converter, mandatory
+            ("name", str, True),
+            ("type", str, True),
+            ("x1", varsep_float, True),
+            ("y1", varsep_float, True),
+            ("x2", varsep_float, True),
+            ("y2", varsep_float, True),
+            ("font", str, False),
+            ("size", varsep_float, False),
+            ("bold", int, False),
+            ("italic", int, False),
+            ("underline", int, False),
+            ("foreground", self._parse_colorcode, False),
+            ("background", self._parse_colorcode, False),
+            ("align", str, False),
+            ("text", str, False),
+            ("priority", int, False),
+            ("multiline", self._parse_multiline, False),
+            ("rotate", varsep_float, False),
         )
         self.elements = []
-        self.pg_no = 0
         if encoding is None:
             encoding = locale.getpreferredencoding()
         with open(infile, encoding=encoding) as f:
             for row in csv.reader(f, delimiter=delimiter):
+                # fill in blanks for any missing items
+                row.extend([""] * (len(key_config) - len(row)))
                 kargs = {}
-                for i, v in enumerate(row):
-                    kargs[handlers[i][0]] = handlers[i][1](v)
+                for val, cfg in zip(row, key_config):
+                    vs = val.strip()
+                    if not vs:
+                        if cfg[2]:  # mandatory
+                            if cfg[0] == "x2" and row["type"] in ["B", "C39"]:
+                                # two types don't need x2, but offset rendering does
+                                pass
+                            else:
+                                raise FPDFException(
+                                    f"Mandatory value '{cfg[0]}' missing in csv data"
+                                )
+                        elif cfg[0] == "priority":
+                            # formally optional, but we need some value for sorting
+                            kargs["priority"] = 0
+                        # otherwise, let the type handlers use their own defaults
+                    else:
+                        kargs[cfg[0]] = cfg[1](vs)
                 self.elements.append(kargs)
-        self.keys = [v["name"].lower() for v in self.elements]
-
-    def add_page(self):
-        self.pg_no += 1
-        self.texts[self.pg_no] = {}
+        self.keys = [val["name"].lower() for val in self.elements]
 
     def __setitem__(self, name, value):
         if name.lower() not in self.keys:
             raise FPDFException(f"Element not loaded, cannot set item: {name}")
-        if not self.pg_no:
-            raise FPDFException("No page open, you need to call add_page() first")
-        self.texts[self.pg_no][name.lower()] = value
+        self.texts[name.lower()] = value
```

Comment:
I wonder why `self.texts` was the only field indexed by "page number" in the original code...
This probably supposed to help rendering documents with several similar pages, with only the text varying from pages to pages.
Anyway, your new API offer more flexibility in that regard.
In there a risk to break backward compatibility here though?

Reply:
`self.texts` holds the only page specific data that the class manages.
The API of Template() hasn't really changed from the user perspective. All existing code that doesn't rely on any internals should still work. In fact, exploiting the internals has become easier now. For example adding content to `self.pdf` from the outside wasn't previously possible (would reccommend FlexTemplate() for that, though).